### PR TITLE
net: tcp: use PSA functions for ISN generation instead of legacy MbedTLS ones

### PR DESCRIPTION
**Pull request by valeriosetti**
This PR:

- replaces MD5 with SHA256 for the computation of ISN (see [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/71747#discussion_r1575859082) for details)
- replaces `mbedtls_xxx()` with `psa_hash_compute()`
- replaces `sys_rand_get()` with `sys_csrand_get()`

This PR is an alternative to #71747 in the sense that there is no `mbedtls_sha256()` alternative option. We go directly to PSA functions instead. This is not a problem because we assume that a PSA provider is always present in the build should it either be built-in into MbedTLS (when `MBEDTLS_PSA_CRYPTO_C` is enabled) or provided externally through TFM (when `BUILD_WITH_TFM` is enabled).

### CODE REVIEW

**Review by jukkar — DISMISSED** [review]
+1 to networking changes.

**Review by rlubos — DISMISSED** [review]

**Review by ceolin — CHANGES_REQUESTED** [review]
It needs clarification about the mix usage of cryptographically safe random generator.
Please check comments.

**Review by ceolin — CHANGES_REQUESTED** [review]
It needs clarification about the mix usage of cryptographically safe random generator.
Please check comments.

File: modules/mbedtls/zephyr_init.c [hunk 1]
```diff
@@ -115,3 +115,33 @@ mbedtls_ms_time_t mbedtls_ms_time(void)
...
+	uint8_t *output, size_t output_size, size_t *output_length)
+{
+	(void) context;
+
+#if defined(CONFIG_CSPRNG_ENABLED)
+	if (sys_csrand_get(output, output_size) != 0) {
+		return PSA_ERROR_GENERIC_ERROR;
+	}
+#else
+	sys_rand_get(output, output_size);
```
[the remark this answers is not in the document]
Comment (valeriosetti) [reply modules/mbedtls/zephyr_init.c]: Yes, sorry, I pushed an updated version that fixes this.

File: subsys/net/ip/Kconfig.tcp [hunk 2]
```diff
@@ -230,8 +230,8 @@ config NET_TCP_ISN_RFC6528
 	default y
 	depends on NET_TCP
 	select MBEDTLS
-	select MBEDTLS_MD
-	select MBEDTLS_MAC_MD5_ENABLED
+	select MBEDTLS_PSA_CRYPTO_C
```
Comment (frkv) [inline subsys/net/ip/Kconfig.tcp:233]:
```suggestion
	select MBEDTLS_PSA_CRYPTO_C depens on !BUILD_WITH_TFM
```
Selecting MBEDTLS_PSA_CRYPTO_C is incompatible with BUILD_WITH_TFM. Not sure if this depends will help, but you could try it out

File: subsys/net/ip/tcp.c [hunk 3]
```diff
@@ -2234,14 +2234,21 @@ static uint32_t tcpv6_init_isn(struct in6_addr *saddr,
 	static bool once;
 
 	if (!once) {
+#if defined(CONFIG_CSPRNG_ENABLED)
+		sys_csrand_get(unique_key, sizeof(unique_key));
+#else
 		sys_rand_get(unique_key, sizeof(unique_key));
+#endif
```
[the remark this answers is not in the document]
Comment (frkv) [reply subsys/net/ip/tcp.c:2298]: Not easy to guess sys_csrand_get doesn't allow test-modes (If I'm not mistaken). If you have predictive RNG numbers then they are definitely guessable. Please use only sys_csrand_get so we can't remove any possible weak holes

**Review by tomi-font — DISMISSED** [review]
Noting that
- some MbedTLS configuration changes come from #71947,
- the non-default enabling of `MBEDTLS_USE_PSA_CRYPTO` if `MBEDTLS_PSA_CRYPTO_C` is temporary (will be possible to revert with #72243),
- and that the newly introduced Kconfig options relating to PSA in `Kconfig.tls-generic` will be moved to `Kconfig.psa` when rebasing this PR.

File: subsys/net/ip/tcp.c [hunk 4]
```diff
@@ -2298,7 +2298,10 @@ static uint32_t tcpv6_init_isn(struct in6_addr *saddr,
 	memcpy(buf.key, unique_key, sizeof(buf.key));
 
 #if defined(CONFIG_NET_TCP_ISN_RFC6528)
-	mbedtls_md5((const unsigned char *)&buf, sizeof(buf), hash);
+	size_t hash_len;
+
+	psa_hash_compute(PSA_ALG_SHA_256, (const unsigned char *)&buf, sizeof(buf),
+					 hash, sizeof(hash), &hash_len);
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2304]: alignment

File: subsys/net/ip/tcp.c [hunk 5]
```diff
@@ -2333,7 +2336,10 @@ static uint32_t tcpv4_init_isn(struct in_addr *saddr,
 	memcpy(buf.key, unique_key, sizeof(unique_key));
 
 #if defined(CONFIG_NET_TCP_ISN_RFC6528)
-	mbedtls_md5((const unsigned char *)&buf, sizeof(buf), hash);
+	size_t hash_len;
+
+	psa_hash_compute(PSA_ALG_SHA_256, (const unsigned char *)&buf, sizeof(buf),
+					 hash, sizeof(hash), &hash_len);
```
Comment (jukkar) [inline subsys/net/ip/tcp.c:2342]: ditto

**Review by frkv — APPROVED** [review]
LGTM (assuming alignment is fixed)

File: subsys/net/ip/tcp.c (hunk 5 above)
Comment (valeriosetti) [reply subsys/net/ip/tcp.c]: Both indentations should be OK now

**Review by jukkar — APPROVED** [review]

**Review by ceolin — APPROVED** [review]

**Pull request merged**